### PR TITLE
✨ Add cloudevents with confluent kafka binding

### PR DESCRIPTION
**Pull request by yanmxa**
## Summary

Bring the cloudevents SDK with confluent kafka binding

## Related issue(s)

Fixes # https://issues.redhat.com/browse/ACM-11002

### CODE REVIEW

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	kafka_confluent "open-cluster-management.io/sdk-go/pkg/cloudevents/generic/options/kafka/protocol"
+	"open-cluster-management.io/sdk-go/pkg/cloudevents/generic/types"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
+)
+
+const (
+	// defaultSpecTopic is a default kafka topic for resource spec.
+	// defaultSpecTopic = "sources/+/clusters/+/spec"
```
Comment (qiujian16) [inline pkg/cloudevents/generic/options/kafka/options.go:13]: not quite clear on the meaning of these comments.

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	defaultSpecTopic         = "spec"
+	defaultStatusTopic       = "status"
+	defaultSpecResyncTopic   = "specresync"
+	defaultStatusResyncTopic = "statusreync"
+)
+
+type KafkaOptions struct {
+	configMap *kafka.ConfigMap
+	Topics    types.Topics
+	// Producer  *kafka.Producer
```
Comment (qiujian16) [inline pkg/cloudevents/generic/options/kafka/options.go:34]: please remove them if we do not need them

File: pkg/cloudevents/generic/options/kafka/protocol/message.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	contentTypeKey = "Content-Type"
+)
+
+const (
+	KafkaOffsetKey    = "kafkaoffset"
+	KafkaPartitionKey = "kafkapartition"
+	KafkaTopicKey     = "kafkatopic"
+	KafkaMessageKey   = "kafkamessagekey"
+)
+
```
Comment (qiujian16) [inline pkg/cloudevents/generic/options/kafka/protocol/message.go:26]: add a comments to describe that this is, I think this is the cloud event implementation, also a TODO to move it to cloudevents/sdk-go with the PR link.

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	// // defaultStatusResyncTopic is a default kafka topic for resource status resync.
+	// defaultStatusResyncTopic = "sources/+/clusters/statusresync"
+
+	defaultSpecTopic         = "spec"
+	defaultStatusTopic       = "status"
+	defaultSpecResyncTopic   = "specresync"
+	defaultStatusResyncTopic = "statusreync"
+)
+
+type KafkaOptions struct {
```
Comment (qiujian16) [inline pkg/cloudevents/generic/options/kafka/options.go:31]: it might be worth adding an example here, it is not clear what is the required setting in the *kafka.ConfigMap

**Review by qiujian16 — COMMENTED** [review]
do we need to add any unit test? there is none right now.

**Comment by yanmxa** [comment]
> do we need to add any unit test? there is none right now.

Yes! I will add the ut later.

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 1 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go]: Remove it

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 2 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go]: Done!

File: pkg/cloudevents/generic/options/kafka/protocol/message.go (hunk 3 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/protocol/message.go:29]: Done!

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 4 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go:22]: I added the test case and also posted the possible link containing the values for the `*kafka.ConfigMap`. https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md

File: pkg/cloudevents/generic/options/kafka/agentoptions.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		kafka_confluent.WithReceiverTopics([]string{o.Topics.Spec, o.Topics.StatusResync}),
+		kafka_confluent.WithSenderTopic(o.Topics.Status),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (o *kafkaAgentOptions) ErrorChan() <-chan error {
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/agentoptions.go:71]: how does this client handle the error, like the reconnect?

**Comment by skeeey** [comment]
we may also need enable this option in the https://github.com/open-cluster-management-io/sdk-go/blob/main/pkg/cloudevents/work/clientbuilder.go#L58

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
+)
+
+const (
+	defaultSpecTopic         = "spec"
+	defaultStatusTopic       = "status"
+	defaultSpecResyncTopic   = "specresync"
+	defaultStatusResyncTopic = "statusreync"
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/options.go:19]: can we remove the reync topics

File: pkg/cloudevents/generic/options/kafka/agentoptions.go (hunk 5 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/agentoptions.go:73]: The confluent kafka recovered automatically if it disconnected. Since we need to sense it. I provide an option to disable the auto recovery and then also define an error chan to deliver the error.

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 6 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go]: Yes, I only keep the "spec" and "status" topic for now.

**Comment by yanmxa** [comment]
> we may also need enable this option in the https://github.com/open-cluster-management-io/sdk-go/blob/main/pkg/cloudevents/work/clientbuilder.go#L58

Done.

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+const (
+	defaultSpecTopic   = "spec"
+	defaultStatusTopic = "status"
+)
+
+type KafkaOptions struct {
+	// the configMap: https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md
+	ConfigMap *kafka.ConfigMap `json:"configs,omitempty" yaml:"configs,omitempty"`
+	Topics    *types.Topics    `json:"topics,omitempty" yaml:"topics,omitempty"`
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/options.go:23]: could we used constant topics (don't expose the topics)? if we allow to set the topics from outside, we may need to define/validate the topic format

Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go]: @skeeey Done!

File: pkg/cloudevents/generic/options/kafka/agentoptions.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+// encode the source and agent to the message key
+func (o *kafkaAgentOptions) WithContext(ctx context.Context, evtCtx cloudevents.EventContext) (context.Context, error) {
+	eventType, err := types.ParseCloudEventsType(evtCtx.GetType())
+	if err != nil {
+		return nil, err
+	}
+
+	topicCtx := cloudeventscontext.WithTopic(ctx, o.Topics.AgentEvents)
+	if eventType.Action == types.ResyncRequestAction {
+		return kafka_confluent.WithMessageKey(topicCtx, o.clusterName), nil
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/agentoptions.go:47]: what does the `MessageKey` mean? is it the topic? how do we use it to do authorize for each cluster?

Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/agentoptions.go:47]: Kafka MessageKey reference to this: https://www.geeksforgeeks.org/apache-kafka-message-keys/

We handle the authorization based on the topic, It has nothing to do with MessageKey.

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	// asterisk is a wildcard for source, the second asterisk is a wildcard for cluster.
+	agentEventsTopic = "agentevents.*.*"
+	// sourceBroadcastTopic is for a source to publish its events to all agents, the asterisk is a wildcard for source.
+	sourceBroadcastTopic = "sourcebroadcast.*"
+	// agentBroadcastTopic is for a agent to publish its events to all sources, the asterisk is a wildcard for cluster.
+	agentBroadcastTopic = "agentbroadcast.*"
+)
+
+type KafkaOptions struct {
+	// TODO don't use this directly
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/options.go:27]: remove these comments

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+				// Generic client instance-level errors, such as
+				// broker connection failures, authentication issues, etc.
+				errChan <- fmt.Errorf("client error %w", ev)
+			}
+		}
+	}()
+}
+
+// BuildKafkaOptionsFromFlags builds configs from a config filepath.
+func BuildKafkaOptionsFromFlags(configPath string) (*KafkaOptions, error) {
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/options.go:72]: how about return the `*kafka.ConfigMap` directly?

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	config := &KafkaOptions{}
+	if err := yaml.Unmarshal(configData, config); err != nil {
+		return nil, err
+	}
+
+	if config.BrokerHost == "" {
+		return nil, fmt.Errorf("brokerHost is required")
+	}
+
+	if config.GroupID == "" {
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/options.go:87]: can we give a default value here?

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 9 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go:27]: Done

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 11 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go:87]: Sure. set it's value as `sourceID` or `agentID`

File: pkg/cloudevents/generic/options/kafka/options.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+	agentEventsTopic = "agentevents.*.*"
+	// sourceBroadcastTopic is for a source to publish its events to all agents, the asterisk is a wildcard for source.
+	sourceBroadcastTopic = "sourcebroadcast.*"
+	// agentBroadcastTopic is for a agent to publish its events to all sources, the asterisk is a wildcard for cluster.
+	agentBroadcastTopic = "agentbroadcast.*"
+)
+
+type KafkaOptions struct {
+	// the configMap: https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md
+	ConfigMap *kafka.ConfigMap `json:"-" yaml:"-"`
```
Comment (skeeey) [inline pkg/cloudevents/generic/options/kafka/options.go:28]: it seems we don't need object in this struct?

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 11 above)
Comment (skeeey) [reply pkg/cloudevents/generic/options/kafka/options.go]:
```
Client group id string. All clients sharing the same group.id belong to the same group.
```
it seems we need a fixed value?

Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go]: Just as we both discussed. the default is agentID for agent and sourceID fro source.

File: pkg/cloudevents/generic/options/kafka/options.go (hunk 12 above)
Comment (yanmxa) [reply pkg/cloudevents/generic/options/kafka/options.go]: Done.

**Comment by skeeey** [comment]
LGTM

**Comment by qiujian16** [comment]
/approve

we should also update the readme here https://github.com/open-cluster-management-io/sdk-go/tree/main/pkg/cloudevents, could be a followup

**Comment by skeeey** [comment]
/lgtm

we may also need to add some integration test for this part as a followup
